File: backend/app/schemas/billing.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from .common import APIModel


class PlanResponse(APIModel):
    code: str
    name: str
    description: str
    price_monthly: float
    status: str


class PlanFeatureResponse(APIModel):
    id: str
    plan_id: str
    feature_key: str
    enabled: bool
    limit_value: int | None = None
    limit_unit: str


class QuotaBalanceResponse(APIModel):
    id: str
    organization_id: str
    quota_key: str
    limit_value: int
    used_value: int
    reserved_value: int
    period_start: datetime
    period_end: datetime
    reset_at: datetime


class UsageEventResponse(APIModel):
    id: str
    organization_id: str
    user_id: str
    project_id: str | None = None
    job_id: str | None = None
    event_type: str
    amount: int
    unit: str
    event_metadata: dict[str, Any] | None = None
    created_at: datetime

File: frontend/types/billing.ts
import type { PlanCode } from "./auth";

export type Plan = {
  code: PlanCode;
  name: string;
  description: string;
  priceMonthly: number;
  priceYearly?: number;
  status: "active" | "archived" | "hidden";
  queuePriority: "low" | "normal" | "high" | "enterprise" | "internal";
  maxConcurrentJobs: number;
  targetUser: string;
};

export type PlanFeature = {
  id: string;
  planCode: PlanCode;
  featureKey: string;
  featureName: string;
  enabled: boolean;
  limitValue: number | "unlimited";
  limitUnit: "words" | "times" | "projects" | "members" | "formats" | "GB" | "level" | "boolean";
  description: string;
};

export type QuotaKey =
  | "monthly_generated_words"
  | "monthly_review_count"
  | "monthly_rewrite_count"
  | "max_projects"
  | "team_members"
  | "concurrent_jobs"
  | "export_docx"
  | "export_epub";

export type QuotaBalance = {
  id: string;
  organizationId: string;
  quotaKey: QuotaKey;
  label: string;
  periodStart: string;
  periodEnd: string;
  limitValue: number;
  usedValue: number;
  reservedValue: number;
  resetAt: string;
};

export type QuotaReservation = {
  id: string;
  organizationId: string;
  generationJobId: string;
  quotaKey: QuotaKey;
  reservedAmount: number;
  consumedAmount: number;
  status: "reserved" | "settled" | "released" | "expired";
  createdAt: string;
  updatedAt: string;
};

export type UsageEvent = {
  id: string;
  organizationId: string;
  userId: string;
  projectId?: string;
  generationJobId?: string;
  eventType: "generate_words" | "review" | "rewrite" | "export" | "manual_adjustment";
  amount: number;
  unit: "words" | "times" | "files";
  createdAt: string;
};
